File: src/main/java/com/xin/server/packet/server/S_LevelDownResult.java
package com.xin.server.packet.server;

import com.xin.server.model.instance.PcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;

/**
 * 境界內降級結果。
 */
public class S_LevelDownResult extends ServerBasePacket {

    public static final String REASON_OK = "OK";

    private S_LevelDownResult(boolean success, String reason, String message) {
        super(ServerOpcodes.S_LEVEL_DOWN_RESULT);
        put("success", success);
        put("reason", reason);
        put("message", message);
    }

    public static S_LevelDownResult ok(PcInstance pc) {
        S_LevelDownResult packet = new S_LevelDownResult(true, REASON_OK, "降級");
        packet.put("realmLevel", pc.getRealmLevel());
        packet.put("exp", pc.getExp());
        packet.put("expMax", pc.getExpMax());
        return packet;
    }
}

File: lib/network/packets/server/s_level_down_result.dart
/// 境界內降級結果。
class SLevelDownResult {
  const SLevelDownResult({
    required this.success,
    required this.reason,
    required this.message,
    required this.realmLevel,
    required this.exp,
    required this.expMax,
  });

  final bool success;
  final String reason;
  final String message;
  final int realmLevel;
  final int exp;
  final int expMax;

  factory SLevelDownResult.fromData(Map<String, dynamic> data) {
    return SLevelDownResult(
      success: data['success'] as bool? ?? false,
      reason: data['reason'] as String? ?? '',
      message: data['message'] as String? ?? '',
      realmLevel: (data['realmLevel'] as num?)?.toInt() ?? 1,
      exp: (data['exp'] as num?)?.toInt() ?? 0,
      expMax: (data['expMax'] as num?)?.toInt() ?? 0,
    );
  }
}
